accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,142 @@
+
+To help the developer know which parts of an application or library the linker is not able to analysis correctly the linker is adding features to warn developers when they are calling linker-unfriendly methods. See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md) for more information.
+
+#### Adding linker annotations
+
+With the addition of the Linker Reflection Flow features above, we will need to add attributes to the .NET Library methods that rely on Reflection. At a high-level, we can run the linker on the libraries and look at the warnings it produces. We then need to analyze the correct fix for the linker warning. Possibilities are:
+
+1. Add a `DynamicallyAccessedMembers` attribute to the code, to tell the linker which members need to be accessed dynamically.
+    - This attribute is similar in nature to the nullable annotations that were added in C# 8 in that they are both viral. `DynamicallyAccessedMembers` needs to be applied from the actual usage of Reflection all the way through the call tree to the public API that a developer calls passing in a Type.
+
+2. In some cases the usage is actually safe. For example, if an "Equals" method is written by retrieving all the fields on a type and looping over them comparing each field. If a field is unused, and therefore trimmed by the linker, there is no need to compare that field for equality. In these cases we can suppress the linker warning using the `[UnconditionalSuppressMessage]` attribute.
+
+3. In some cases the API itself is not able to be linker friendly. In these cases we don't want to suppress the warning completely, but instead pass the warning to the caller of the API. In these cases we can use the `[RequiresUnreferencedCode]` attribute. This attribute tells the linker to not warn about this method, but instead warn about any callers of this method.
+
+A high-level goal is that the .NET Libraries are clean of linker warnings. This is one level of confidence we can have that .NET applications will be linker safe.
+
+#### Serialization
+
+An especially hard situation for the linker to understand is code that uses Reflection to (de)serialize object graphs to and from some external representation. Examples in the .NET Libraries include (but are not limited to):
+
+- `XmlSerializer`
+- `JsonSerializer`
+- `EventSource.Write<T>()`
+
+One reason this is hard is that every serialization mechanism has its own rules about what it looks for:
```

Comment:
I would specifically mention "serialization and deserialization" - for example I often think about "serialization" as the way to write objects to text. I don't make the connection that most people mean also "deserialization" which is the opposite. And the problems are specific to serialize/deserialize (.ctor problem is deserialize only for example).

Reply:
fixed.